docs/okeps/okep-4456-egressip-on-user-defined-networks.md

```diff
@@ -0,0 +1,315 @@
+
+#### Story 2
+
+As a user of OVN-Kubernetes, I should be able to utilize the EgressIP high availability feature for primary user defined L2/L3 networks.
+
+#### Story 3
+
+As a user of OVN-Kubernetes, I should be able to utilize the EgressIP feature for pods which only have connectivity to the cluster default network.
+
+#### Story 4
+
+As a user of OVN-Kubernetes, I should be able to continue EgressIP on pods where some sets of pods have a cluster default network as their primary network while other sets of pods have user defined networks as their primary network.
+
+#### Story 5
+
+As a user of OVN-Kubernetes, I should be able to utilize the EgressIP feature to allow consistent source IP when egress-ing secondary host interfaces from pods which have connectivity to primary user defined networks.
+
+## Proposed Solution
+
+The EgressIP should work seamlessly irrespective of whether the pod is either managed by a primary cluster default network or primary user defined networks or overlapping subnets being used in user defined networks across namespaces, the same EgressIP is used to manage different user defined networks, etc. From a user perspective this should not have any impact on EgressIP at API level. At the same time this poses some challenges for EgressIP in the dataplane to send and receive pod egress traffic.
+
+Consider the scenario, podA and podB assigned with same IP address 10.244.0.8, but assigned from different user defined networks (say from netA and netB),
+both pods are served by same EgressIP 1.1.1.1, now two pods are trying to connect externally using the same source and destination port, when SNAT happens on OVN's Gateway Router (GR), both pod traffic would have same 5-tuple and collision happen at br-ex. To avoid this, SNAT must be moved from OVN gateway router into br-ex  with appropriate flow rules to avoid the collision.
+
+When traffic from both EgressIP and Non-EgressIP pods arrive at br-ex with GR masquerade IP used as the source IP address, there should be a way to distinguish which packets require SNAT for EgressIP, otherwise it would be SNATed with the node's IP address. The EgressIP pod is packet marked using OVN's Logical Router Policy (LRP) and flow rule in br-ex match on `pkt_mark` to apply SNAT.
+
+### Packet marks
+We do not require a packet mark per IP within each EgressIP CR because only one IP is assigned per node per EgressIP CR.
+Packet mark allocation must not conflict with other features using packet marks, be within the specific integer range defined by
+the linux kernel and also not conflict with other features utilizing packet marks.
+
+There isn't a reliable way for each node controller to generate a deterministic packet mark based on the constraints mentioned previously
+for each EgressIP CR.
+
+Packet marks (integer) will be centrally allocated by a cluster manager managed allocator and annotated on each EgressIP CR.
```

Comment:
note to self: check to see upgrade order - if node controllers updates before CM, they could be blocked reconfiguring EIP during upgrade until CM is upgraded.

Reply:
Initially, we thought we could derive the mark from the EIP CR name or UID but thats a bad idea for a number of reasons (name can change, string needs to map to int within a range and it may clash).
So, we thought CM seems a good place to allocate it but during upgrade, EIP config would be blocked until CM upgrades. Need input.